### PR TITLE
Automate creation of Issue for major release PHP synchronisation

### PR DESCRIPTION
**Pull request by getdave**
## What?

Automates the creation of the contents for the Issue that tracks PHP changes ([example](https://github.com/WordPress/gutenberg/issues/54177)) that may need to be synchronised to WP Core for a major release.

## Why?

For a major release, the Editor Tech Lead needs to create an Issue like [this one](https://github.com/WordPress/gutenberg/issues/54177) which details all the PHP changes that may need backport to Core.

This requires manually going through all the key directories in the project that may contain PHP changes and then checking all the PRs that have been merged since the final RC of the _previous_ release. If any PRs contain PHP changes then the PR needs to be added to the Issue in a very specific format. 

It also requires organising headings for each seciton and doing a lot of manual formatting.

All this is:

- laborious - I'm told it typically requires a full day or more to complete
- prone to error - it's easy to make a typo or miss a change

This is why I think it's wide open for automation.

## How?

This PR has a script which can be run locally which will

- get all commits to `trunk` in a number of key directories (including `/lib`, `/block-library` and `/phpunit`) since a date you provide.
- get the details for each of the commits including the files that were changed
- put the commit under the correct heading based on which files were changed
- dedupe the result set
- output the markdown required to create the Issue

I've made some subjective choices about how things are group and displayed:

- ~PRs may be listed only once in the Issue - the exception is PRs which modify PHP Unit tests as I wanted to display a full list of changes that affect the `/phpunit` directory.~ Actually PRs are allowed more than once as long as it's across different sections
- No more than 3 level of nesting - anything beyond that gets very visually messy. I've grouped PRs that would appear under those subheadings under the third level of nesting.
- Certain directories are excluded including `e2e-tests` and `experiments-page.php` which are never synchronised - please feel free to suggest additions to that list.
- I've elected to avoid the additional network requests required to fetch the details of the PRs associated with each commit. Instead I've used regex to attempt to build the PR URL from the commit message. This isn't 100% reliable but it works in the majority of cases and we have a reasonable fallback in place.

## Testing Instructions

- Get [GH Personal Access Token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token)
- Run `node bin/generate-php-sync-issue.mjs --token={YOUR_TOKEN_HERE} --since=2023-11-01`
- Open `issueContent.md` and copy all contents.
- Go to https://github.com/WordPress/gutenberg/issues/new/ and paste contents.
- `Preview` the Issue - **DO NOT SUBMIT IT!**
- Take any given PR in the Issue and check that the files it modified match up with where it is listed in the Issue.

### CODE REVIEW

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs:190]: Currently writes to a file. We could have it automatically create the Issue on Github.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs]: This can generate a fair number of requests

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs]: Not using `path` will result in a tonne of commits that have little or no value. So actually it's better to only fetch commits against the paths we know will have PHP files that need to be synced.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs:256]: If we can't build the PR URL from the commit message, then we'll fall back to the commit URL itself and let GH UI show the link to the PR there.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs]:
```suggestion
	return `- [ ] ${ pr.url } - @${ pr.creator } | Trac ticket | Core backport PR\n`;
```

We'll need to commit this before we merge this PR.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs]: We should probably remove this or use a dedicated flag for "low-network" or something/

**Comment by Mamaduka** [comment]
I believe @ramonjd was working on something similar last year.

File: bin/generate-php-sync-issue.mjs
Comment (Mamaduka) [inline bin/generate-php-sync-issue.mjs]: The `packages/block-library` is automatically synced.

Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: I noticed it was included in https://github.com/WordPress/gutenberg/issues/54177.

If we want to exclude we can add it to the ignore list and remove from paths.

Comment (Mamaduka) [reply bin/generate-php-sync-issue.mjs]: There's no manual action required for syncing block library changes. We can safely remove it.

Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: ✅

**Comment by getdave** [comment]
Manually checking this against PRs in `/lib` has resulted in the following missing PRs:

- https://github.com/WordPress/gutenberg/pull/57639/

In addition the following PRs missed changes in lib but were "caught" in the PHPUnit section:

- #57866

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs]: There is a bug here where the individual commits are being removed from the data rather than moved to their parent.

Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: I decided to remove this for now. If we want to restore the feature to remove single-PR levels then we'll probably need to adjust the data structure to be:

```
{
'lib': {
commits: [], // any commits for this level
children: {} // more leels here
}
```

That will afford a means to move any PRs listed under a level which only has a single commit listed to instead be placed on the parent level's commits.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs:41]: If folks agree with the premise then some tests for this function would be helpful.

File: bin/generate-php-sync-issue.mjs
Comment (tellthemachines) [reply bin/generate-php-sync-issue.mjs]: Should be fine to just commit it? Nobody will actually be pinged until the issue is created, right?

File: bin/generate-php-sync-issue.mjs
Comment (tellthemachines) [inline bin/generate-php-sync-issue.mjs:107]: This is just merge commits to trunk, right?

File: bin/generate-php-sync-issue.mjs
Comment (tellthemachines) [reply bin/generate-php-sync-issue.mjs:190]: I'm not sure about this, since the generated content will almost certainly need a bit of editing. Might just be easier to write to file and the person generating it can copy/paste/edit before publishing.

File: bin/generate-php-sync-issue.mjs
Comment (tellthemachines) [inline bin/generate-php-sync-issue.mjs]: This'll need to be a bit bigger to accommodate the length of WP release cycles 😅

**Review by tellthemachines — COMMENTED** [review]
Thanks for the PR, this is working pretty well for me! 

Regarding the generated content:

* the `compat/wordpress-6.4` folder can be removed, since whatever is in there is already in core. It exists to provide compatibility for WP 6.3, because we have to support the current and the previous WP versions. Not sure if there's a simple way to automate that as the folder numbers will vary per release - might be easier to just delete it manually.
* `load.php` is plugin-specific so we probably don't need to look at it either.
* not sure if this is possible: could we remove some PRs from the list if they have a certain label? Anything with `Backport from WordPress Core`  can safely be ignored.
* Would it be worth checking the `Needs PHP backport` label and including any PRs from there that aren't in the list?

One thing to bear in mind is that we should include all PRs since the last plugin release to be fully included in Beta 1, because anytime after that new features could have been merged that would still need syncing. See my comment on `validateSince` below 😄

Other than that I think this is looking pretty good! I'm no node expert though so you might want a second opinion on the code itself.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs:107]: It uses https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#list-commits and the `path` arg to get the commits.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs:190]: Sounds fine to me. It's a very manual script anyway. The overhead of auto-creating an Issue probably isn't worth the effort.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: What would you suggest? 4?

You have to provide the `--since` arg anyway (which @mikachan said should be set to date of last RC of previous major release) so this is just a safe guard to ensure that if someone accidentally provides a date far in the past we're not hitting up the API for thousands of commits 😓

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: I was testing the output by creating a new Issue and pasting in the code. A couple of times my muscle memory nearly ended up submitting the Issue. So this is a safeguard 😄

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [inline bin/generate-php-sync-issue.mjs]:
```suggestion
```

We can ditch this.

File: bin/generate-php-sync-issue.mjs
Comment (youknowriad) [reply bin/generate-php-sync-issue.mjs:190]: I agree we shouldn't create an issue but we could actually output the content in the console. So the commands become something like `yourcommand > issueContent.md` if you really want to put the content in a file. That seems more inline with how tools like that work.

**Review by youknowriad — COMMENTED** [review]
This looks great to me. Some valid remarks above from @tellthemachines but otherwise it's a perfect head start to the backporting process. Nice work everyone.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: > include all PRs since the last plugin release to be fully included in Beta 1

You said this. So for this release, according to [this doc](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/versions-in-wordpress.md) that would be [16.7](https://github.com/WordPress/gutenberg/releases/tag/v16.7.0) which was 2023-09-23.

Am I right?

**Comment by getdave** [comment]
Thanks for review @tellthemachines 👍 

> * the `compat/wordpress-6.4` folder can be removed, since whatever is in there is already in core. It exists to provide compatibility for WP 6.3, because we have to support the current and the previous WP versions. Not sure if there's a simple way to automate that as the folder numbers will vary per release - might be easier to just delete it manually.

Currently it will remove the folder for the current stable release but I think we need to make it a whitelist to retain only the folder for the _next_ release and remove all others.


> * `load.php` is plugin-specific so we probably don't need to look at it either.

✅ Done

> * not sure if this is possible: could we remove some PRs from the list if they have a certain label? Anything with `Backport from WordPress Core`  can safely be ignored.

Apparently this label gets removed once the PR has been backported as a way of indicating it's completion status. So instead @youknowriad and I agreed that we would flag any PRs made between Beta 1 and final RC with a special warning.

> * Would it be worth checking the `Needs PHP backport` label and including any PRs from there that aren't in the list?

Yes that would be a nice addition.

> One thing to bear in mind is that we should include all PRs since the last plugin release to be fully included in Beta 1, because anytime after that new features could have been merged that would still need syncing. See my comment on `validateSince` below 😄

`--since` will work with whatever you provide. I've extended the max allowed date.

**Comment by getdave** [comment]
Given that this is not production code we could probably look to merge and iterate on this.

File: bin/generate-php-sync-issue.mjs
Comment (mikachan) [inline bin/generate-php-sync-issue.mjs]: I think I got this wrong when we discussed this earlier 🤦‍♀️ It should be all changes from the final version of Gutenberg that's bundled with the last WordPress release. So for 6.5, it would be any changes from GB 16.8 onwards, as 16.7 was the last version of the plugin included in 6.4. There's a table with the included version numbers [here](https://developer.wordpress.org/block-editor/contributors/versions-in-wordpress/).

There will be some overlap from bugs fixes, etc but this should capture all changes. Sorry for the confusion!

**Comment by youknowriad** [comment]
> not sure if this is possible: could we remove some PRs from the list if they have a certain label? Anything with Backport from WordPress Core can safely be ignored.

@getdave "from WordPress Core" is another label and these can be safely ignored regardless of date...

**Comment by ramonjd** [comment]
> I believe @ramonjd was working on something similar last year.

Thanks for the ping. I had a local shell script trying to do what this PR's script does, but mine was definitely not as good and detailed as this one, and required too much manual clean up. Thanks for this initiative @getdave!

File: bin/generate-php-sync-issue.mjs
Comment (tellthemachines) [reply bin/generate-php-sync-issue.mjs]: Yes! 4 months should be ok.

File: bin/generate-php-sync-issue.mjs
Comment (tellthemachines) [inline bin/generate-php-sync-issue.mjs]: Typo: Backport.

> In future releases we will instead retain the label and add a **new** label

I think adding a new label is a great idea so we can identify the backported PRs and rule them out, but we still need to remove `Backport to WP beta/RC` because we need to be able to filter PRs by that label and retrieve only the ones that haven't been done yet.

**Review by tellthemachines — APPROVED** [review]
LGTM! Small non-blocking typo and comment below. I think we can merge and iterate.

**Review by Mamaduka — APPROVED** [review]
Excellent work, Dave.

Maybe, as a follow-up, we can document the usage of this script. Here is an example of cherry-picking automation docs - https://developer.wordpress.org/block-editor/contributors/code/release/auto-cherry-picking/.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: Ah yes. I'll remember that. Thank you.

**Comment by getdave** [comment]
> * Backport from WordPress Core

✅ I've set up an ignore list for any labels. This can be added to in followups.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: ✅ Updated.

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs]: ✅ Updated

File: bin/generate-php-sync-issue.mjs
Comment (getdave) [reply bin/generate-php-sync-issue.mjs:107]: Confirmed this defaults to `trunk`.

**Comment by getdave** [comment]
Just noting I used this script to generate https://github.com/WordPress/gutenberg/issues/57959

**Pull request merged**
